### PR TITLE
all: add arbitrary capability support

### PR DESCRIPTION
**Pull request by bradfitz**
Updates #4217

### CODE REVIEW

File: wgengine/filter/filter.go [hunk 1]
```diff
@@ -205,6 +214,27 @@ func matchesFamily(ms matches, keep func(netaddr.IP) bool) matches {
 	return ret
 }
 
+// capMatchesFamily returns the subset of ms for which keep(srcNet.IP)
+// and the match is a capability grant.
+func capMatchesFamily(ms matches, keep func(netaddr.IP) bool) matches {
```
Comment (maisem) [inline wgengine/filter/filter.go:219]: `capMatchFamily` implies that it is an inherent understanding of IP families. Maybe rename this and `matchesFamily` to `matchesFunc`. Or leave it as is ¯\\\_(ツ)\_/¯

File: wgengine/filter/filter.go [hunk 2]
```diff
@@ -291,6 +321,30 @@ func (f *Filter) CheckTCP(srcIP, dstIP netaddr.IP, dstPort uint16) Response {
 	return f.RunIn(pkt, 0)
 }
 
+// AppendCaps appends to base the capabilities that srcIP has talking
+// to dstIP.
+func (f *Filter) AppendCaps(base []string, srcIP, dstIP netaddr.IP) []string {
```
Comment (maisem) [inline wgengine/filter/filter.go:326]: optional: this is doing too much, maybe split out `Caps`?

File: wgengine/filter/filter.go (hunk 1 above)
Comment (maisem) [reply wgengine/filter/filter.go:219]: why not match on `cap.Net` too, to be consistent with `matchesFamily`?

File: wgengine/filter/match.go [hunk 3]
```diff
@@ -47,12 +47,18 @@ func (npr NetPortRange) String() string {
 	return fmt.Sprintf("%v:%v", npr.Net, npr.Ports)
 }
 
+type CapMatch struct {
+	Net netaddr.IPPrefix
+	Cap string
+}
```
Comment (maisem) [inline wgengine/filter/match.go:53]: docs

File: wgengine/filter/match.go [hunk 4]
```diff
@@ -47,12 +47,18 @@ func (npr NetPortRange) String() string {
 	return fmt.Sprintf("%v:%v", npr.Net, npr.Ports)
 }
 
+type CapMatch struct {
+	Net netaddr.IPPrefix
```
Comment (maisem) [inline wgengine/filter/match.go:51]: rename Net to Dst?

File: wgengine/filter/match.go [hunk 5]
```diff
@@ -47,12 +47,18 @@ func (npr NetPortRange) String() string {
 	return fmt.Sprintf("%v:%v", npr.Net, npr.Ports)
 }
 
+type CapMatch struct {
```
Comment (maisem) [inline wgengine/filter/match.go:50]: This name reuse really threw me off in the code review and I had to open it in VSCode to actually figure out what was going on.

File: wgengine/filter/match.go (hunk 4 above)
Comment (bradfitz) [reply wgengine/filter/match.go:51]: done

File: wgengine/filter/match.go (hunk 3 above)
Comment (bradfitz) [reply wgengine/filter/match.go:53]: done

**Review by maisem — APPROVED** [review]

**Pull request merged**